# Assignment 3
# Author: Grismely M. Hiraldo

In [20]:
import numpy as np
import matplotlib.pyplot as plt 
from mpl_toolkits import mplot3d
from sklearn import datasets
from sklearn.linear_model import Perceptron
from sklearn.linear_model import SGDClassifier
from sklearn.linear_model import LogisticRegression
from sklearn import svm
from sklearn.svm import LinearSVC
from sklearn.pipeline import make_pipeline
from sklearn.preprocessing import StandardScaler
from sklearn.datasets import make_classification

In [3]:
# import the Iris data set

iris = datasets.load_iris()
X = iris.data
Y = iris.target
Xzeros0 = (0. * X) + .2
Xzeros1 = (0. * X) 
Xzeros2 = (0. * X) - .2
x_min, x_max = X.min() - 0.5, X.max() + 1.5

In [4]:
# get Iris classifications

IC = np.unique(Y)
IrisC = ('Setosa','Versicolor','Virginica')
pltC = ('red', 'blue', 'green')
pltCback = ('pink', 'lightblue','lightgreen')
pltM = ('o', 'x', '*')
for i in range(0,3):
    print (IC[i],pltM[i],IrisC[i],pltC[i],pltCback[i])

0 o Setosa red pink
1 x Versicolor blue lightblue
2 * Virginica green lightgreen


# Stochastic Gradient Descent using the modified-Huber loss function

In [5]:
# choose the classes and features to be included in the model

cdx = np.array([0,1,2])             # this array contains the classes to be considered (3 classes)
fdx = np.array([0,1])           # this array contains the features to be considered ( AB features)

nC = len(cdx)
nF = len(fdx)

# build the XX and YY arrays
# these arrays contain the samples in the classes specified in cdx and include the features specified in fdx

i = 0
numY = 0
for i in range(len(Y)):
    for j in range(nC):
        if Y[i] == IC[cdx[j]]:
            numY += 1

YY = np.zeros((numY))
XX = np.zeros((numY,nF))
if (len(fdx) == 1):
    XXzeros0 = (0. * XX) + .2
    XXzeros1 = (0. * XX) 
    XXzeros2 = (0. * XX) - .2

j = 0
for i in range(len(Y)):
    for k in range(nC):
        if Y[i] == IC[cdx[k]]:        
            YY[j] = IC[cdx[k]]
            for m in range(nF):
                XX[j,m] = X[i,fdx[m]] 
            j += 1
            
# fit the model

# clf = Perceptron(tol=1e-4, random_state=0)
clf = SGDClassifier(tol=1e-4, random_state=0)

clf.fit(XX,YY)

print('\nScore: ',clf.score(XX,YY))



Score:  0.7733333333333333


The accuracy of the model on the subset of the data used for this pair is only .773, which can be intereppted as the model being semi accurate on this particular combination of features but because of previous experience (assignment 1) with this model I anticipate to find more a more accurate model on the data even on all the subet of feature combinations.

In [6]:
# choose the classes and features to be included in the model

cdx = np.array([0,1,2])             # this array contains the classes to be considered (3 classes)
fdx = np.array([0,2])           # this array contains the features to be considered ( AC features)

nC = len(cdx)
nF = len(fdx)

# build the XX and YY arrays
# these arrays contain the samples in the classes specified in cdx and include the features specified in fdx

i = 0
numY = 0
for i in range(len(Y)):
    for j in range(nC):
        if Y[i] == IC[cdx[j]]:
            numY += 1

YY = np.zeros((numY))
XX = np.zeros((numY,nF))
if (len(fdx) == 1):
    XXzeros0 = (0. * XX) + .2
    XXzeros1 = (0. * XX) 
    XXzeros2 = (0. * XX) - .2

j = 0
for i in range(len(Y)):
    for k in range(nC):
        if Y[i] == IC[cdx[k]]:        
            YY[j] = IC[cdx[k]]
            for m in range(nF):
                XX[j,m] = X[i,fdx[m]] 
            j += 1
            
# fit the model

# clf = Perceptron(tol=1e-4, random_state=0)
clf = SGDClassifier(tol=1e-4, random_state=0)

clf.fit(XX,YY)

print('\nScore: ',clf.score(XX,YY))



Score:  0.8733333333333333


The model has a higher accuracy on this subset of features by 10% this could be because its been observed previously that features can affect accuracy.

In [7]:
# choose the classes and features to be included in the model

cdx = np.array([0,1,2])             # this array contains the classes to be considered (3 classes)
fdx = np.array([0,3])           # this array contains the features to be considered ( AD features)

nC = len(cdx)
nF = len(fdx)

# build the XX and YY arrays
# these arrays contain the samples in the classes specified in cdx and include the features specified in fdx

i = 0
numY = 0
for i in range(len(Y)):
    for j in range(nC):
        if Y[i] == IC[cdx[j]]:
            numY += 1

YY = np.zeros((numY))
XX = np.zeros((numY,nF))
if (len(fdx) == 1):
    XXzeros0 = (0. * XX) + .2
    XXzeros1 = (0. * XX) 
    XXzeros2 = (0. * XX) - .2

j = 0
for i in range(len(Y)):
    for k in range(nC):
        if Y[i] == IC[cdx[k]]:        
            YY[j] = IC[cdx[k]]
            for m in range(nF):
                XX[j,m] = X[i,fdx[m]] 
            j += 1
            
# fit the model

# clf = Perceptron(tol=1e-4, random_state=0)
clf = SGDClassifier(tol=1e-4, random_state=0)

clf.fit(XX,YY)

print('\nScore: ',clf.score(XX,YY))



Score:  0.94


The highest score thus far for the model.

In [8]:
# choose the classes and features to be included in the model

cdx = np.array([0,1,2])             # this array contains the classes to be considered (3 classes)
fdx = np.array([1,2])           # this array contains the features to be considered ( BC features)

nC = len(cdx)
nF = len(fdx)

# build the XX and YY arrays
# these arrays contain the samples in the classes specified in cdx and include the features specified in fdx

i = 0
numY = 0
for i in range(len(Y)):
    for j in range(nC):
        if Y[i] == IC[cdx[j]]:
            numY += 1

YY = np.zeros((numY))
XX = np.zeros((numY,nF))
if (len(fdx) == 1):
    XXzeros0 = (0. * XX) + .2
    XXzeros1 = (0. * XX) 
    XXzeros2 = (0. * XX) - .2

j = 0
for i in range(len(Y)):
    for k in range(nC):
        if Y[i] == IC[cdx[k]]:        
            YY[j] = IC[cdx[k]]
            for m in range(nF):
                XX[j,m] = X[i,fdx[m]] 
            j += 1
            
# fit the model

# clf = Perceptron(tol=1e-4, random_state=0)
clf = SGDClassifier(tol=1e-4, random_state=0)

clf.fit(XX,YY)

print('\nScore: ',clf.score(XX,YY))



Score:  0.8866666666666667


A relavtiely high score of accuracy for the model again not 100% but high enough that we cannot cross out the model as bad or inaccurate.

In [58]:
# choose the classes and features to be included in the model

cdx = np.array([0,1,2])             # this array contains the classes to be considered (3 classes)
fdx = np.array([1,3])           # this array contains the features to be considered ( BD features)

nC = len(cdx)
nF = len(fdx)

# build the XX and YY arrays
# these arrays contain the samples in the classes specified in cdx and include the features specified in fdx

i = 0
numY = 0
for i in range(len(Y)):
    for j in range(nC):
        if Y[i] == IC[cdx[j]]:
            numY += 1

YY = np.zeros((numY))
XX = np.zeros((numY,nF))
if (len(fdx) == 1):
    XXzeros0 = (0. * XX) + .2
    XXzeros1 = (0. * XX) 
    XXzeros2 = (0. * XX) - .2

j = 0
for i in range(len(Y)):
    for k in range(nC):
        if Y[i] == IC[cdx[k]]:        
            YY[j] = IC[cdx[k]]
            for m in range(nF):
                XX[j,m] = X[i,fdx[m]] 
            j += 1
            
# fit the model

# clf = Perceptron(tol=1e-4, random_state=0)
clf = SGDClassifier(tol=1e-4, random_state=0)

clf.fit(XX,YY)

print('\nScore: ',clf.score(XX,YY))



Score:  0.8933333333333333


Starting to see more consistency in the scores for this model

In [10]:
# choose the classes and features to be included in the model

cdx = np.array([0,1,2])             # this array contains the classes to be considered (3 classes)
fdx = np.array([2,3])           # this array contains the features to be considered ( CD features)

nC = len(cdx)
nF = len(fdx)

# build the XX and YY arrays
# these arrays contain the samples in the classes specified in cdx and include the features specified in fdx

i = 0
numY = 0
for i in range(len(Y)):
    for j in range(nC):
        if Y[i] == IC[cdx[j]]:
            numY += 1

YY = np.zeros((numY))
XX = np.zeros((numY,nF))
if (len(fdx) == 1):
    XXzeros0 = (0. * XX) + .2
    XXzeros1 = (0. * XX) 
    XXzeros2 = (0. * XX) - .2

j = 0
for i in range(len(Y)):
    for k in range(nC):
        if Y[i] == IC[cdx[k]]:        
            YY[j] = IC[cdx[k]]
            for m in range(nF):
                XX[j,m] = X[i,fdx[m]] 
            j += 1
            
# fit the model

# clf = Perceptron(tol=1e-4, random_state=0)
clf = SGDClassifier(tol=1e-4, random_state=0)

clf.fit(XX,YY)

print('\nScore: ',clf.score(XX,YY))



Score:  0.96


The highest accuracy score the model had. Overall, the score was about 0.885 or 88.5 with a low outlier of 0.77 and some high 90s. I think it can be said that the model was decent it could still be considered accurate as it did have, with the exception of the .77, mid 90s and high 80 scores because the scores did flux I would say that while the model was decent we can say confidently that it is probably not the best model to train for our data and we could find better.

# Logistic regression

In [11]:
# choose the classes and features to be included in the model

cdx = np.array([0,1,2])             # this array contains the classes to be considered (3 classes)
fdx = np.array([0,1])           # this array contains the features to be considered ( AB features)

nC = len(cdx)
nF = len(fdx)

# build the XX and YY arrays
# these arrays contain the samples in the classes specified in cdx and include the features specified in fdx

i = 0
numY = 0
for i in range(len(Y)):
    for j in range(nC):
        if Y[i] == IC[cdx[j]]:
            numY += 1

YY = np.zeros((numY))
XX = np.zeros((numY,nF))
if (len(fdx) == 1):
    XXzeros0 = (0. * XX) + .2
    XXzeros1 = (0. * XX) 
    XXzeros2 = (0. * XX) - .2

j = 0
for i in range(len(Y)):
    for k in range(nC):
        if Y[i] == IC[cdx[k]]:        
            YY[j] = IC[cdx[k]]
            for m in range(nF):
                XX[j,m] = X[i,fdx[m]] 
            j += 1
            
# fit the model

# clf = Perceptron(tol=1e-4, random_state=0)
log_reg = LogisticRegression()
log_reg.fit(XX,YY)

print('\nScore: ',log_reg.score(XX,YY))



Score:  0.9666666666666667


The accuracy of the model on the subset of the data used for this pair is 0.96 it is nearly 100% accurate which indicates that the model was trained well over this subset and is/will produce accurate results.

In [12]:
# choose the classes and features to be included in the model

cdx = np.array([0,1,2])             # this array contains the classes to be considered (3 classes)
fdx = np.array([0,2])           # this array contains the features to be considered ( AC features)

nC = len(cdx)
nF = len(fdx)

# build the XX and YY arrays
# these arrays contain the samples in the classes specified in cdx and include the features specified in fdx

i = 0
numY = 0
for i in range(len(Y)):
    for j in range(nC):
        if Y[i] == IC[cdx[j]]:
            numY += 1

YY = np.zeros((numY))
XX = np.zeros((numY,nF))
if (len(fdx) == 1):
    XXzeros0 = (0. * XX) + .2
    XXzeros1 = (0. * XX) 
    XXzeros2 = (0. * XX) - .2

j = 0
for i in range(len(Y)):
    for k in range(nC):
        if Y[i] == IC[cdx[k]]:        
            YY[j] = IC[cdx[k]]
            for m in range(nF):
                XX[j,m] = X[i,fdx[m]] 
            j += 1
            
# fit the model

# clf = Perceptron(tol=1e-4, random_state=0)
log_reg = LogisticRegression()
log_reg.fit(XX,YY)

print('\nScore: ',log_reg.score(XX,YY))



Score:  0.96


Same accuracy score as the last subset, there already seems to be more consistency with the logistic regression model which I think is very befitting because even a general log function is not linear and tends to encomposs and show trends for more irregular data while linear functions can often leave out clusters of data like outliers. I think this model would benefit from plotting.

In [13]:
# choose the classes and features to be included in the model

cdx = np.array([0,1,2])             # this array contains the classes to be considered (3 classes)
fdx = np.array([0,3])           # this array contains the features to be considered ( AD features)

nC = len(cdx)
nF = len(fdx)

# build the XX and YY arrays
# these arrays contain the samples in the classes specified in cdx and include the features specified in fdx

i = 0
numY = 0
for i in range(len(Y)):
    for j in range(nC):
        if Y[i] == IC[cdx[j]]:
            numY += 1

YY = np.zeros((numY))
XX = np.zeros((numY,nF))
if (len(fdx) == 1):
    XXzeros0 = (0. * XX) + .2
    XXzeros1 = (0. * XX) 
    XXzeros2 = (0. * XX) - .2

j = 0
for i in range(len(Y)):
    for k in range(nC):
        if Y[i] == IC[cdx[k]]:        
            YY[j] = IC[cdx[k]]
            for m in range(nF):
                XX[j,m] = X[i,fdx[m]] 
            j += 1
            
# fit the model

# clf = Perceptron(tol=1e-4, random_state=0)
log_reg = LogisticRegression()
log_reg.fit(XX,YY)

print('\nScore: ',log_reg.score(XX,YY))



Score:  0.96


The model continues to show consitnecy and accuracy with a strong score of 96%.

In [59]:


# choose the classes and features to be included in the model

cdx = np.array([0,1,2])             # this array contains the classes to be considered (3 classes)
fdx = np.array([1,2])           # this array contains the features to be considered ( BC features)

nC = len(cdx)
nF = len(fdx)

# build the XX and YY arrays
# these arrays contain the samples in the classes specified in cdx and include the features specified in fdx

i = 0
numY = 0
for i in range(len(Y)):
    for j in range(nC):
        if Y[i] == IC[cdx[j]]:
            numY += 1

YY = np.zeros((numY))
XX = np.zeros((numY,nF))
if (len(fdx) == 1):
    XXzeros0 = (0. * XX) + .2
    XXzeros1 = (0. * XX) 
    XXzeros2 = (0. * XX) - .2

j = 0
for i in range(len(Y)):
    for k in range(nC):
        if Y[i] == IC[cdx[k]]:        
            YY[j] = IC[cdx[k]]
            for m in range(nF):
                XX[j,m] = X[i,fdx[m]] 
            j += 1
            
# fit the model

# clf = Perceptron(tol=1e-4, random_state=0)
log_reg = LogisticRegression()
log_reg.fit(XX,YY)

print('\nScore: ',log_reg.score(XX,YY))



Score:  0.9533333333333334


The model continues to show consitnecy and accuracy.

In [15]:
# choose the classes and features to be included in the model

cdx = np.array([0,1,2])             # this array contains the classes to be considered (3 classes)
fdx = np.array([1,3])           # this array contains the features to be considered ( BD features)

nC = len(cdx)
nF = len(fdx)

# build the XX and YY arrays
# these arrays contain the samples in the classes specified in cdx and include the features specified in fdx

i = 0
numY = 0
for i in range(len(Y)):
    for j in range(nC):
        if Y[i] == IC[cdx[j]]:
            numY += 1

YY = np.zeros((numY))
XX = np.zeros((numY,nF))
if (len(fdx) == 1):
    XXzeros0 = (0. * XX) + .2
    XXzeros1 = (0. * XX) 
    XXzeros2 = (0. * XX) - .2

j = 0
for i in range(len(Y)):
    for k in range(nC):
        if Y[i] == IC[cdx[k]]:        
            YY[j] = IC[cdx[k]]
            for m in range(nF):
                XX[j,m] = X[i,fdx[m]] 
            j += 1
            
# fit the model

# clf = Perceptron(tol=1e-4, random_state=0)
log_reg = LogisticRegression()
log_reg.fit(XX,YY)

print('\nScore: ',log_reg.score(XX,YY))



Score:  0.96


The model continues to show consitnecy and accuracy with a another score of 96%.

In [16]:
# choose the classes and features to be included in the model

cdx = np.array([0,1,2])             # this array contains the classes to be considered (3 classes)
fdx = np.array([2,3])           # this array contains the features to be considered (CD features)

nC = len(cdx)
nF = len(fdx)

# build the XX and YY arrays
# these arrays contain the samples in the classes specified in cdx and include the features specified in fdx

i = 0
numY = 0
for i in range(len(Y)):
    for j in range(nC):
        if Y[i] == IC[cdx[j]]:
            numY += 1

YY = np.zeros((numY))
XX = np.zeros((numY,nF))
if (len(fdx) == 1):
    XXzeros0 = (0. * XX) + .2
    XXzeros1 = (0. * XX) 
    XXzeros2 = (0. * XX) - .2

j = 0
for i in range(len(Y)):
    for k in range(nC):
        if Y[i] == IC[cdx[k]]:        
            YY[j] = IC[cdx[k]]
            for m in range(nF):
                XX[j,m] = X[i,fdx[m]] 
            j += 1
            
# fit the model

# clf = Perceptron(tol=1e-4, random_state=0)
log_reg = LogisticRegression()
log_reg.fit(XX,YY)

print('\nScore: ',log_reg.score(XX,YY))



Score:  0.9666666666666667


Overall, it seems that logistic regression was trained well over the data and overall is a good fit. Although the model did not produce an 100% or 1.0 score there was consistency in the accuracy scores as they all were approxiamantly 96 with the exception of one that was 95, and the scores themselves are also quite high.

# Support Vector Machine - Linear

In [42]:
# choose the classes and features to be included in the model

cdx = np.array([0,1,2])             # this array contains the classes to be considered (3 classes)
fdx = np.array([0,1])           # this array contains the features to be considered (AB features)

nC = len(cdx)
nF = len(fdx)

# build the XX and YY arrays
# these arrays contain the samples in the classes specified in cdx and include the features specified in fdx

i = 0
numY = 0
for i in range(len(Y)):
    for j in range(nC):
        if Y[i] == IC[cdx[j]]:
            numY += 1

YY = np.zeros((numY))
XX = np.zeros((numY,nF))
if (len(fdx) == 1):
    XXzeros0 = (0. * XX) + .2
    XXzeros1 = (0. * XX) 
    XXzeros2 = (0. * XX) - .2

j = 0
for i in range(len(Y)):
    for k in range(nC):
        if Y[i] == IC[cdx[k]]:        
            YY[j] = IC[cdx[k]]
            for m in range(nF):
                XX[j,m] = X[i,fdx[m]] 
            j += 1
            
# fit the model
linear_svc = svm.SVC(kernel='linear')
linear_svc.fit(XX,YY);
print('\nScore: ',linear_svc.score(XX,YY))


Score:  0.82


The linear SVM model over the data had a score of 0.82, a pretty good score but it should be noted that sepal length and width the pair the model is used on for this set has been observed to usually result in lower accuracy in the past, this may just be a result of the model not training well over these features but if it continues to occur from the previous assignment to this one it may be that there is a very large disperaity in the data or that it is vastly scattered so the model is training with many outliters and very little patterns.

In [39]:
# choose the classes and features to be included in the model

cdx = np.array([0,1,2])             # this array contains the classes to be considered (3 classes)
fdx = np.array([0,2])           # this array contains the features to be considered (AC features)

nC = len(cdx)
nF = len(fdx)

# build the XX and YY arrays
# these arrays contain the samples in the classes specified in cdx and include the features specified in fdx

i = 0
numY = 0
for i in range(len(Y)):
    for j in range(nC):
        if Y[i] == IC[cdx[j]]:
            numY += 1

YY = np.zeros((numY))
XX = np.zeros((numY,nF))
if (len(fdx) == 1):
    XXzeros0 = (0. * XX) + .2
    XXzeros1 = (0. * XX) 
    XXzeros2 = (0. * XX) - .2

j = 0
for i in range(len(Y)):
    for k in range(nC):
        if Y[i] == IC[cdx[k]]:        
            YY[j] = IC[cdx[k]]
            for m in range(nF):
                XX[j,m] = X[i,fdx[m]] 
            j += 1
            
# fit the model

linear_svc = svm.SVC(kernel='linear')
linear_svc.fit(XX,YY);
print('\nScore: ',linear_svc.score(XX,YY))


Score:  0.9533333333333334


The model has 95% accuracy over this pair.

In [36]:
# choose the classes and features to be included in the model

cdx = np.array([0,1,2])             # this array contains the classes to be considered (3 classes)
fdx = np.array([0,3])           # this array contains the features to be considered (AD features)

nC = len(cdx)
nF = len(fdx)

# build the XX and YY arrays
# these arrays contain the samples in the classes specified in cdx and include the features specified in fdx

i = 0
numY = 0
for i in range(len(Y)):
    for j in range(nC):
        if Y[i] == IC[cdx[j]]:
            numY += 1

YY = np.zeros((numY))
XX = np.zeros((numY,nF))
if (len(fdx) == 1):
    XXzeros0 = (0. * XX) + .2
    XXzeros1 = (0. * XX) 
    XXzeros2 = (0. * XX) - .2

j = 0
for i in range(len(Y)):
    for k in range(nC):
        if Y[i] == IC[cdx[k]]:        
            YY[j] = IC[cdx[k]]
            for m in range(nF):
                XX[j,m] = X[i,fdx[m]] 
            j += 1
            
# fit the model

linear_svc = svm.SVC(kernel='linear')
linear_svc.fit(XX,YY);
print('\nScore: ',linear_svc.score(XX,YY))


Score:  0.96


There is some consistency between this pair and the previous.

In [35]:
# choose the classes and features to be included in the model

cdx = np.array([0,1,2])             # this array contains the classes to be considered (3 classes)
fdx = np.array([1,2])           # this array contains the features to be considered (BC features)

nC = len(cdx)
nF = len(fdx)

# build the XX and YY arrays
# these arrays contain the samples in the classes specified in cdx and include the features specified in fdx

i = 0
numY = 0
for i in range(len(Y)):
    for j in range(nC):
        if Y[i] == IC[cdx[j]]:
            numY += 1

YY = np.zeros((numY))
XX = np.zeros((numY,nF))
if (len(fdx) == 1):
    XXzeros0 = (0. * XX) + .2
    XXzeros1 = (0. * XX) 
    XXzeros2 = (0. * XX) - .2

j = 0
for i in range(len(Y)):
    for k in range(nC):
        if Y[i] == IC[cdx[k]]:        
            YY[j] = IC[cdx[k]]
            for m in range(nF):
                XX[j,m] = X[i,fdx[m]] 
            j += 1
            
# fit the model

linear_svc = svm.SVC(kernel='linear')
linear_svc.fit(XX,YY);
print('\nScore: ',linear_svc.score(XX,YY))


Score:  0.96


The model is showing consistency and accuracy with a recuring score of 96%.

In [34]:
# choose the classes and features to be included in the model

cdx = np.array([0,1,2])             # this array contains the classes to be considered (3 classes)
fdx = np.array([1,3])           # this array contains the features to be considered (BD features)

nC = len(cdx)
nF = len(fdx)

# build the XX and YY arrays
# these arrays contain the samples in the classes specified in cdx and include the features specified in fdx

i = 0
numY = 0
for i in range(len(Y)):
    for j in range(nC):
        if Y[i] == IC[cdx[j]]:
            numY += 1

YY = np.zeros((numY))
XX = np.zeros((numY,nF))
if (len(fdx) == 1):
    XXzeros0 = (0. * XX) + .2
    XXzeros1 = (0. * XX) 
    XXzeros2 = (0. * XX) - .2

j = 0
for i in range(len(Y)):
    for k in range(nC):
        if Y[i] == IC[cdx[k]]:        
            YY[j] = IC[cdx[k]]
            for m in range(nF):
                XX[j,m] = X[i,fdx[m]] 
            j += 1
            
# fit the model

linear_svc = svm.SVC(kernel='linear')
linear_svc.fit(XX,YY);
print('\nScore: ',linear_svc.score(XX,YY))


Score:  0.96


The model is showing consistency and accuracy with a recuring score of 96%.

In [33]:
# choose the classes and features to be included in the model

cdx = np.array([0,1,2])             # this array contains the classes to be considered (3 classes)
fdx = np.array([2,3])           # this array contains the features to be considered (CD features)

nC = len(cdx)
nF = len(fdx)

# build the XX and YY arrays
# these arrays contain the samples in the classes specified in cdx and include the features specified in fdx

i = 0
numY = 0
for i in range(len(Y)):
    for j in range(nC):
        if Y[i] == IC[cdx[j]]:
            numY += 1

YY = np.zeros((numY))
XX = np.zeros((numY,nF))
if (len(fdx) == 1):
    XXzeros0 = (0. * XX) + .2
    XXzeros1 = (0. * XX) 
    XXzeros2 = (0. * XX) - .2

j = 0
for i in range(len(Y)):
    for k in range(nC):
        if Y[i] == IC[cdx[k]]:        
            YY[j] = IC[cdx[k]]
            for m in range(nF):
                XX[j,m] = X[i,fdx[m]] 
            j += 1
            
# fit the model

linear_svc = svm.SVC(kernel='linear')
linear_svc.fit(XX,YY);
print('\nScore: ',linear_svc.score(XX,YY))


Score:  0.9666666666666667


This model also seemed to have trained well over the data with just one outlier of 88% which is still a decently high score. This was honestly to my surprise as mentioned previously sometimes linear functions and models can exclude clusters of data. I think we could also cofnidently say this is a good model for our data as it showed accuracy and conisstency through all subsets of data.

# Support Vector Machine - Polynomial

In [51]:
# choose the classes and features to be included in the model

cdx = np.array([0,1,2])             # this array contains the classes to be considered (3 classes)
fdx = np.array([0,1])           # this array contains the features to be considered (AB features)

nC = len(cdx)
nF = len(fdx)

# build the XX and YY arrays
# these arrays contain the samples in the classes specified in cdx and include the features specified in fdx

i = 0
numY = 0
for i in range(len(Y)):
    for j in range(nC):
        if Y[i] == IC[cdx[j]]:
            numY += 1

YY = np.zeros((numY))
XX = np.zeros((numY,nF))
if (len(fdx) == 1):
    XXzeros0 = (0. * XX) + .2
    XXzeros1 = (0. * XX) 
    XXzeros2 = (0. * XX) - .2

j = 0
for i in range(len(Y)):
    for k in range(nC):
        if Y[i] == IC[cdx[k]]:        
            YY[j] = IC[cdx[k]]
            for m in range(nF):
                XX[j,m] = X[i,fdx[m]] 
            j += 1
            
# fit the model

poly_svc = svm.SVC(kernel='poly')
poly_svc.fit(XX,YY);
print('\nScore: ',poly_svc.score(XX,YY))


Score:  0.8133333333333334


The point made about sepal width and length seems to hold true across all models, although the score is still pretty good it is always the lowest score between all six subsets, this tells me that one of my two predictions about the data may be true as we expose the data set to more models it seems more and more true that the models are underfitting the subset pair of sepal width and length or that the data itself may have discrepenscies.

In [46]:
# choose the classes and features to be included in the model

cdx = np.array([0,1,2])             # this array contains the classes to be considered (3 classes)
fdx = np.array([0,2])           # this array contains the features to be considered (AC features)

nC = len(cdx)
nF = len(fdx)

# build the XX and YY arrays
# these arrays contain the samples in the classes specified in cdx and include the features specified in fdx

i = 0
numY = 0
for i in range(len(Y)):
    for j in range(nC):
        if Y[i] == IC[cdx[j]]:
            numY += 1

YY = np.zeros((numY))
XX = np.zeros((numY,nF))
if (len(fdx) == 1):
    XXzeros0 = (0. * XX) + .2
    XXzeros1 = (0. * XX) 
    XXzeros2 = (0. * XX) - .2

j = 0
for i in range(len(Y)):
    for k in range(nC):
        if Y[i] == IC[cdx[k]]:        
            YY[j] = IC[cdx[k]]
            for m in range(nF):
                XX[j,m] = X[i,fdx[m]] 
            j += 1
            
# fit the model

poly_svc = svm.SVC(kernel='poly')
poly_svc.fit(XX,YY);
print('\nScore: ',poly_svc.score(XX,YY))


Score:  0.96


The model has 96% accuracy over this pair.

In [47]:
# choose the classes and features to be included in the model

cdx = np.array([0,1,2])             # this array contains the classes to be considered (3 classes)
fdx = np.array([0,3])           # this array contains the features to be considered (AD features)

nC = len(cdx)
nF = len(fdx)

# build the XX and YY arrays
# these arrays contain the samples in the classes specified in cdx and include the features specified in fdx

i = 0
numY = 0
for i in range(len(Y)):
    for j in range(nC):
        if Y[i] == IC[cdx[j]]:
            numY += 1

YY = np.zeros((numY))
XX = np.zeros((numY,nF))
if (len(fdx) == 1):
    XXzeros0 = (0. * XX) + .2
    XXzeros1 = (0. * XX) 
    XXzeros2 = (0. * XX) - .2

j = 0
for i in range(len(Y)):
    for k in range(nC):
        if Y[i] == IC[cdx[k]]:        
            YY[j] = IC[cdx[k]]
            for m in range(nF):
                XX[j,m] = X[i,fdx[m]] 
            j += 1
            
# fit the model

poly_svc = svm.SVC(kernel='poly')
poly_svc.fit(XX,YY);
print('\nScore: ',poly_svc.score(XX,YY))


Score:  0.9533333333333334


The model has 95% accuracy over this pair, after our initial pair we are starting to see consistency in the scores just as the previous model.

In [48]:
# choose the classes and features to be included in the model

cdx = np.array([0,1,2])             # this array contains the classes to be considered (3 classes)
fdx = np.array([1,2])           # this array contains the features to be considered (BC features)

nC = len(cdx)
nF = len(fdx)

# build the XX and YY arrays
# these arrays contain the samples in the classes specified in cdx and include the features specified in fdx

i = 0
numY = 0
for i in range(len(Y)):
    for j in range(nC):
        if Y[i] == IC[cdx[j]]:
            numY += 1

YY = np.zeros((numY))
XX = np.zeros((numY,nF))
if (len(fdx) == 1):
    XXzeros0 = (0. * XX) + .2
    XXzeros1 = (0. * XX) 
    XXzeros2 = (0. * XX) - .2

j = 0
for i in range(len(Y)):
    for k in range(nC):
        if Y[i] == IC[cdx[k]]:        
            YY[j] = IC[cdx[k]]
            for m in range(nF):
                XX[j,m] = X[i,fdx[m]] 
            j += 1
            
# fit the model

poly_svc = svm.SVC(kernel='poly')
poly_svc.fit(XX,YY);
print('\nScore: ',poly_svc.score(XX,YY))


Score:  0.9533333333333334


The model has 95% accuracy over this pair.

In [49]:
# choose the classes and features to be included in the model

cdx = np.array([0,1,2])             # this array contains the classes to be considered (3 classes)
fdx = np.array([1,3])           # this array contains the features to be considered (BD features)

nC = len(cdx)
nF = len(fdx)

# build the XX and YY arrays
# these arrays contain the samples in the classes specified in cdx and include the features specified in fdx

i = 0
numY = 0
for i in range(len(Y)):
    for j in range(nC):
        if Y[i] == IC[cdx[j]]:
            numY += 1

YY = np.zeros((numY))
XX = np.zeros((numY,nF))
if (len(fdx) == 1):
    XXzeros0 = (0. * XX) + .2
    XXzeros1 = (0. * XX) 
    XXzeros2 = (0. * XX) - .2

j = 0
for i in range(len(Y)):
    for k in range(nC):
        if Y[i] == IC[cdx[k]]:        
            YY[j] = IC[cdx[k]]
            for m in range(nF):
                XX[j,m] = X[i,fdx[m]] 
            j += 1
            
# fit the model

poly_svc = svm.SVC(kernel='poly')
poly_svc.fit(XX,YY);
print('\nScore: ',poly_svc.score(XX,YY))


Score:  0.96


The model has 96% accuracy over this pair.

In [50]:
# choose the classes and features to be included in the model

cdx = np.array([0,1,2])             # this array contains the classes to be considered (3 classes)
fdx = np.array([2,3])           # this array contains the features to be considered (CD features)

nC = len(cdx)
nF = len(fdx)

# build the XX and YY arrays
# these arrays contain the samples in the classes specified in cdx and include the features specified in fdx

i = 0
numY = 0
for i in range(len(Y)):
    for j in range(nC):
        if Y[i] == IC[cdx[j]]:
            numY += 1

YY = np.zeros((numY))
XX = np.zeros((numY,nF))
if (len(fdx) == 1):
    XXzeros0 = (0. * XX) + .2
    XXzeros1 = (0. * XX) 
    XXzeros2 = (0. * XX) - .2

j = 0
for i in range(len(Y)):
    for k in range(nC):
        if Y[i] == IC[cdx[k]]:        
            YY[j] = IC[cdx[k]]
            for m in range(nF):
                XX[j,m] = X[i,fdx[m]] 
            j += 1
            
# fit the model

poly_svc = svm.SVC(kernel='poly')
poly_svc.fit(XX,YY);
print('\nScore: ',poly_svc.score(XX,YY))


Score:  0.9666666666666667


The polynomial SVM model was very similar to the linear SVM model over our data, the first subset (sepal width, sepal length) has the lowest accuracy score for both and after there it assumes more consistency and accuracy with a recurging score of 95 or 96 just as the previous model did. Since this is the second use of SVM models we could potentially say with the next model as proof that the way the SVM model trains over sepal width and length is resulting in underfitting and may be why in both SVM models that subset has the lowest accuracy despite the model showing consistency and accuracy over the rest of the data.

# Support Vector Machine - RBF

In [52]:
# choose the classes and features to be included in the model

cdx = np.array([0,1,2])             # this array contains the classes to be considered (3 classes)
fdx = np.array([0,1])           # this array contains the features to be considered (AB features)

nC = len(cdx)
nF = len(fdx)

# build the XX and YY arrays
# these arrays contain the samples in the classes specified in cdx and include the features specified in fdx

i = 0
numY = 0
for i in range(len(Y)):
    for j in range(nC):
        if Y[i] == IC[cdx[j]]:
            numY += 1

YY = np.zeros((numY))
XX = np.zeros((numY,nF))
if (len(fdx) == 1):
    XXzeros0 = (0. * XX) + .2
    XXzeros1 = (0. * XX) 
    XXzeros2 = (0. * XX) - .2

j = 0
for i in range(len(Y)):
    for k in range(nC):
        if Y[i] == IC[cdx[k]]:        
            YY[j] = IC[cdx[k]]
            for m in range(nF):
                XX[j,m] = X[i,fdx[m]] 
            j += 1
            
# fit the model

rbf_svc = svm.SVC(kernel='rbf')
rbf_svc.fit(XX,YY);
print('\nScore: ',rbf_svc.score(XX,YY))


Score:  0.82


The accuracy of the model over this subset is 82% this supports our previous claims that the SVM models could be underfitting this particular subset as for the third time this subset has the lowest accuracy score, unlike the logistic regression that had no outlier.

In [53]:
# choose the classes and features to be included in the model

cdx = np.array([0,1,2])             # this array contains the classes to be considered (3 classes)
fdx = np.array([0,2])           # this array contains the features to be considered (AC features)

nC = len(cdx)
nF = len(fdx)

# build the XX and YY arrays
# these arrays contain the samples in the classes specified in cdx and include the features specified in fdx

i = 0
numY = 0
for i in range(len(Y)):
    for j in range(nC):
        if Y[i] == IC[cdx[j]]:
            numY += 1

YY = np.zeros((numY))
XX = np.zeros((numY,nF))
if (len(fdx) == 1):
    XXzeros0 = (0. * XX) + .2
    XXzeros1 = (0. * XX) 
    XXzeros2 = (0. * XX) - .2

j = 0
for i in range(len(Y)):
    for k in range(nC):
        if Y[i] == IC[cdx[k]]:        
            YY[j] = IC[cdx[k]]
            for m in range(nF):
                XX[j,m] = X[i,fdx[m]] 
            j += 1
            
# fit the model

rbf_svc = svm.SVC(kernel='rbf')
rbf_svc.fit(XX,YY);
print('\nScore: ',rbf_svc.score(XX,YY))


Score:  0.96


The model has 96% accuracy over this pair.

In [54]:
# choose the classes and features to be included in the model

cdx = np.array([0,1,2])             # this array contains the classes to be considered (3 classes)
fdx = np.array([0,3])           # this array contains the features to be considered (AD features)

nC = len(cdx)
nF = len(fdx)

# build the XX and YY arrays
# these arrays contain the samples in the classes specified in cdx and include the features specified in fdx

i = 0
numY = 0
for i in range(len(Y)):
    for j in range(nC):
        if Y[i] == IC[cdx[j]]:
            numY += 1

YY = np.zeros((numY))
XX = np.zeros((numY,nF))
if (len(fdx) == 1):
    XXzeros0 = (0. * XX) + .2
    XXzeros1 = (0. * XX) 
    XXzeros2 = (0. * XX) - .2

j = 0
for i in range(len(Y)):
    for k in range(nC):
        if Y[i] == IC[cdx[k]]:        
            YY[j] = IC[cdx[k]]
            for m in range(nF):
                XX[j,m] = X[i,fdx[m]] 
            j += 1
            
# fit the model

rbf_svc = svm.SVC(kernel='rbf')
rbf_svc.fit(XX,YY);
print('\nScore: ',rbf_svc.score(XX,YY))


Score:  0.96


The model has 95% accuracy over this pair, second occurance of the score for this model, we will now see a consitency in it just like the models before.

In [55]:
# choose the classes and features to be included in the model

cdx = np.array([0,1,2])             # this array contains the classes to be considered (3 classes)
fdx = np.array([1,2])           # this array contains the features to be considered (BC features)

nC = len(cdx)
nF = len(fdx)

# build the XX and YY arrays
# these arrays contain the samples in the classes specified in cdx and include the features specified in fdx

i = 0
numY = 0
for i in range(len(Y)):
    for j in range(nC):
        if Y[i] == IC[cdx[j]]:
            numY += 1

YY = np.zeros((numY))
XX = np.zeros((numY,nF))
if (len(fdx) == 1):
    XXzeros0 = (0. * XX) + .2
    XXzeros1 = (0. * XX) 
    XXzeros2 = (0. * XX) - .2

j = 0
for i in range(len(Y)):
    for k in range(nC):
        if Y[i] == IC[cdx[k]]:        
            YY[j] = IC[cdx[k]]
            for m in range(nF):
                XX[j,m] = X[i,fdx[m]] 
            j += 1
            
# fit the model

rbf_svc = svm.SVC(kernel='rbf')
rbf_svc.fit(XX,YY);
print('\nScore: ',rbf_svc.score(XX,YY))


Score:  0.9533333333333334


The model has 95% accuracy over this pair.

In [61]:
# choose the classes and features to be included in the model

cdx = np.array([0,1,2])             # this array contains the classes to be considered (3 classes)
fdx = np.array([1,3])           # this array contains the features to be considered (BD features)

nC = len(cdx)
nF = len(fdx)

# build the XX and YY arrays
# these arrays contain the samples in the classes specified in cdx and include the features specified in fdx

i = 0
numY = 0
for i in range(len(Y)):
    for j in range(nC):
        if Y[i] == IC[cdx[j]]:
            numY += 1

YY = np.zeros((numY))
XX = np.zeros((numY,nF))
if (len(fdx) == 1):
    XXzeros0 = (0. * XX) + .2
    XXzeros1 = (0. * XX) 
    XXzeros2 = (0. * XX) - .2

j = 0
for i in range(len(Y)):
    for k in range(nC):
        if Y[i] == IC[cdx[k]]:        
            YY[j] = IC[cdx[k]]
            for m in range(nF):
                XX[j,m] = X[i,fdx[m]] 
            j += 1
            
# fit the model

rbf_svc = svm.SVC(kernel='rbf')
rbf_svc.fit(XX,YY);
print('\nScore: ',rbf_svc.score(XX,YY))


Score:  0.96


The model has 96% accuracy over this pair.

In [57]:
# choose the classes and features to be included in the model

cdx = np.array([0,1,2])             # this array contains the classes to be considered (3 classes)
fdx = np.array([2,3])           # this array contains the features to be considered (CD features)

nC = len(cdx)
nF = len(fdx)

# build the XX and YY arrays
# these arrays contain the samples in the classes specified in cdx and include the features specified in fdx

i = 0
numY = 0
for i in range(len(Y)):
    for j in range(nC):
        if Y[i] == IC[cdx[j]]:
            numY += 1

YY = np.zeros((numY))
XX = np.zeros((numY,nF))
if (len(fdx) == 1):
    XXzeros0 = (0. * XX) + .2
    XXzeros1 = (0. * XX) 
    XXzeros2 = (0. * XX) - .2

j = 0
for i in range(len(Y)):
    for k in range(nC):
        if Y[i] == IC[cdx[k]]:        
            YY[j] = IC[cdx[k]]
            for m in range(nF):
                XX[j,m] = X[i,fdx[m]] 
            j += 1
            
# fit the model

rbf_svc = svm.SVC(kernel='rbf')
rbf_svc.fit(XX,YY);
print('\nScore: ',rbf_svc.score(XX,YY))


Score:  0.9533333333333334


Not unlike the other SVM models after the first subset there was more consistency and accuracy in the 95-96% range but unlike its predecessors the rbf model did result in more 95% scores which is not a drastic difference but should still be noted so that we may properly critique and compare our models.

# Conclusion

It was very interesting and refreshing to use different models over familar data. It can be said, naturally, that some models were trained better over the data and were a better or more accuracte fit. It was also interesting to see previous notes and assumptions I had made in our last assignment hold true or be brought up again. I will use this to open my discussion about the models used in this assignment. Once again there seems to be something off, seemingly underfitting with the subset data pair (Sepal length and Sepal width) across all models except logistic regression. The models that seem to be underfited over this subset do not have low accuracy scores however their lowest accuracy scores all are from this particular pair. This leads me to presume there is soemthing off in how the model was trained over it or that there may be some discrepency in the actual data but the later seems more possible. Moving on, the first model, the Stochastic gradient descent using the modified-Huber loss function, performed as expected. Like it did previously it was good but not great there were certain scores that told us that there was a better fit for our data. The Logistic Regression Model I believed performed the best and is the best fit for our data, it was consistent and accurate and the only exception to our sepal length and width dilemma. Although it did not result in a score of 100% it approach it at 96% over ever subset pair, the only model to do so. The SVM models are performed similarly, the first subset pair always resulted in a mid 80s accuracy score and then there would be recuring 96s and 95s after. Overall, I don't think we can rule any model as bad or truly inaccurate they all performed pretty similar, Logistic Regression can be ruled as simply the better fit because its was consitent through all the subsets (6/6) while most were (5/6). The last thing I think it is important to make note of is that none of our models resutled in 100% over our data.I would not say this calls for concer because I speculate that it stems from the fact our models were only being trained over subset pairs and not all four features. I tested this by adding all four features (0,1,2,3) to three of our models and the accuracy scores went from 96 to 98. So despite no 100% accuracy on our models they are still good models and I would be interested to see how all the models do with the four features included.